pkg/globalnet/controllers/node_controller.go

```diff
@@ -0,0 +1,220 @@
+*/
+package controllers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/federate"
+	"github.com/submariner-io/admiral/pkg/syncer"
+	admUtil "github.com/submariner-io/admiral/pkg/util"
+	"github.com/submariner-io/submariner/pkg/globalnet/controllers/iptables"
+	"github.com/submariner-io/submariner/pkg/ipam"
+	"github.com/submariner-io/submariner/pkg/routeagent_driver/constants"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+)
+
+func NewNodeController(config syncer.ResourceSyncerConfig, pool *ipam.IPPool, nodeName string) (Interface, error) {
+	var err error
+
+	klog.Info("Creating Node controller")
+
+	iptIface, err := iptables.New()
+	if err != nil {
+		return nil, errors.WithMessage(err, "error creating the IPTablesInterface handler")
+	}
+
+	controller := &nodeController{
+		baseIPAllocationController: newBaseIPAllocationController(pool, iptIface),
+	}
+
+	controller.nodeName = nodeName
+	controller.resourceSyncer, err = syncer.NewResourceSyncer(&syncer.ResourceSyncerConfig{
+		Name:                "Node syncer",
+		ResourceType:        &corev1.Node{},
+		SourceClient:        config.SourceClient,
+		SourceNamespace:     corev1.NamespaceAll,
+		RestMapper:          config.RestMapper,
+		Federator:           federate.NewUpdateFederator(config.SourceClient, config.RestMapper, corev1.NamespaceAll),
+		Scheme:              config.Scheme,
+		Transform:           controller.process,
+		ResourcesEquivalent: controller.onNodeUpdated,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	_, gvr, err := admUtil.ToUnstructuredResource(&corev1.Node{}, config.RestMapper)
+	if err != nil {
+		return nil, err
+	}
+
+	federator := federate.NewUpdateFederator(config.SourceClient, config.RestMapper, corev1.NamespaceAll)
+	controller.nodes = config.SourceClient.Resource(*gvr)
+	localNodeInfo, err := controller.nodes.Get(context.TODO(), controller.nodeName, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.WithMessagef(err, "error retrieving localNodeInfo for %q", controller.nodeName)
+	}
+
+	if localNodeInfo != nil {
+		if err := controller.reserveAllocatedIPs(federator, localNodeInfo); err != nil {
+			return nil, err
+		}
+	}
+
+	return controller, nil
+}
+
+func (n *nodeController) process(from runtime.Object, numRequeues int, op syncer.Operation) (runtime.Object, bool) {
+	node := from.(*corev1.Node)
+
+	// If the event corresponds to a different node which has globalIP annotation, release the globalIP back to Pool.
+	if node.Name != n.nodeName {
+		if existingGlobalIP := node.GetAnnotations()[SmGlobalIP]; existingGlobalIP != "" {
+			if op == syncer.Delete {
+				_ = n.pool.Release(existingGlobalIP)
+				return nil, false
+			} else {
+				_ = n.pool.Release(existingGlobalIP)
+				return n.updateNodeAnnotation(node, "")
+			}
+		}
+
+		return nil, false
+	}
+
```

Comment:
What about delete here? I guess that can't happen  as this process would no longer be running if its node was deleted?

Reply:
Yup, exactly.